Workspaces list

chiragsalian opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
$ https://github.com/Expensify/App/issues/9971

### Tests / QA Steps
1. Create a new account. Login, click on profile icon to view settings.
2. Confirm you see a new "Workspaces" option above "Profile".
3. Click into it and confirm you see "Create a new workspace" message with text explaining about a workspace below it. Confirm you see a button for "New Workspace" at the bottom.
4. Click on the New Workspace button. Confirm a workspace is created successfully.
5. Click back to go to the Settings and confirm you see 1 workspace avatars next to Workspaces.
6. Repeat steps 4 and 5, until you have 4 workspaces and confirm you see 4 avatars.
7. Repeat steps 4 and 5 again for a 5th workspace, this time confirm you see 4 avatars but the 4rth one is opague with +1 over it.
8. Add workspace profile icons for the 1st, 3rd and 4rth workspaces. Go to settings and confirm their avatars show up with the proper icons uploaded. (the 4th one has opacity so you might have to zoom in to see clearer)
9. Click on workspaces -> select a workspace -> manage members -> set network throttling to offline in chrome dev tools -> invite user `+25857527444` (this is an incorrect user and should throw error). Close settings. 
10. Turn network throttling to online. Confirm you see error red dot on profile icon, click it.
11. Confirm you see error red dot next to the avatars and the UI still looks good.

- [x] Verify that no errors appear in the JS console

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The reviewer will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web

https://user-images.githubusercontent.com/14356145/197073921-e3e6a6c1-8b93-4956-b9b6-cb537931d693.mov


#### Mobile Web - Chrome
![image](https://user-images.githubusercontent.com/14356145/197082261-d291a545-d01a-464d-a5f3-ac4d56d30e59.png)

![image](https://user-images.githubusercontent.com/14356145/197082299-e5763e0e-2947-4098-a6cd-15474a448a51.png)

![image](https://user-images.githubusercontent.com/14356145/197082348-facf024a-cd48-4e11-98f5-3ba3eb05048d.png)

![image](https://user-images.githubusercontent.com/14356145/197082385-0f6d7ff2-00c1-488c-a8f6-fd75855fad22.png)

#### Mobile Web - Safari
![image](https://user-images.githubusercontent.com/14356145/197069488-d1cb81e1-2461-49a8-9b46-5003ad75a2be.png)

![image](https://user-images.githubusercontent.com/14356145/197069515-bffa0cb2-8b43-4259-a062-d9c2e1b9c9a2.png)

![image](https://user-images.githubusercontent.com/14356145/197069593-e0d809fb-3f80-4b49-841a-1a6c31b8e8e0.png)

![image](https://user-images.githubusercontent.com/14356145/197069998-b4eef18e-0393-4cac-80af-185ee3d77402.png)

#### Desktop
![image](https://user-images.githubusercontent.com/14356145/197074864-c4680e9e-3993-4e58-b7eb-6fd4b61de0dd.png)

![image](https://user-images.githubusercontent.com/14356145/197074890-5673d8b7-cd79-4f65-99a5-b9986afa9426.png)

![image](https://user-images.githubusercontent.com/14356145/197074957-52a7d189-8fe7-4127-930d-988b29893197.png)

![image](https://user-images.githubusercontent.com/14356145/197075024-e43feb7a-6846-4e12-bb9f-e6ef3f9df6b4.png)


#### iOS
![image](https://user-images.githubusercontent.com/14356145/197064807-86d4085e-1e8e-43b3-9985-993067b3b0c0.png)

![image](https://user-images.githubusercontent.com/14356145/197064834-5c6ec898-e119-4623-b007-64d083d803ef.png)

![image](https://user-images.githubusercontent.com/14356145/197069402-04c0d2c3-ef84-4749-9c43-6788269ced6d.png)

![image](https://user-images.githubusercontent.com/14356145/197070048-9710034c-4d52-4b79-aa7e-619e206a5e25.png)


#### Android
![image](https://user-images.githubusercontent.com/14356145/197080798-f3e9b6c5-9a5f-4150-907a-7db3953d5780.png)

![image](https://user-images.githubusercontent.com/14356145/197080819-a09a979d-1721-452b-bdc0-772c667f58ae.png)

![image](https://user-images.githubusercontent.com/14356145/197080888-a010b6ff-7faa-45f9-9f64-95e8f0089616.png)

![image](https://user-images.githubusercontent.com/14356145/197080928-9985f2a9-2631-4bcf-b1c6-3f99fb6b4af1.png)

chiragsalian committed:
Adding workspaces to languages

chiragsalian committed:
a new route and menu item

chiragsalian committed:
a new modal screen

chiragsalian committed:
updating modal stack navigator with new workspacesPage

chiragsalian committed:
New workspaces page

chiragsalian committed:
correcting the back button

chiragsalian committed:
Path rename so all workspace files are in the same place

chiragsalian committed:
removing unused code and trying some stuff

chiragsalian committed:
adding avatars to the end of workspace

chiragsalian committed:
support for fallbackIcons. unsure why avatar component is not used instead of Image

chiragsalian committed:
passing policy avatars

chiragsalian committed:
Removing new workspace button from 3 dots and adding it as footer

chiragsalian committed:
defining AVATAR_SIZE SMALLER

chiragsalian committed:
cleaning up MultipleAvatars component to work with both avatar and icon

chiragsalian committed:
shouldStackHorizontally param

chiragsalian committed:
styles

chiragsalian committed:
displaying multiple avatars vertically

chiragsalian committed:
using building icon instead of workspace

chiragsalian committed:
update to show both avatar and overlay

chiragsalian committed:
overlay opacity and style fixes

chiragsalian committed:
style cleanup

chiragsalian committed:
cleanup

chiragsalian committed:
perfecting the styles

chiragsalian committed:
style corrections to make the avatars work when there are less than 4 icons

chiragsalian committed:
Merge branch 'main' into chirag-workspaces-list

chiragsalian committed:
adding a positioning style

chiragsalian committed:
correcting avatar styles

chiragsalian committed:
correct avatar styles with red brick road indicator

chiragsalian committed:
adding blocking view for when workspaces are empty

chiragsalian committed:
workspaces language update

chiragsalian committed:
correcting a style

chiragsalian committed:
Merge branch 'main' into chirag-workspaces-list

chiragsalian committed:
logic correction

chiragsalian committed:
cleanup

chiragsalian committed:
Merge branch 'main' into chirag-workspaces-list

chiragsalian committed:
copy corrections

chiragsalian committed:
cleanup

chiragsalian committed:
added key to fix js console error

chiragsalian committed:
updating avatar to match order

chiragsalian committed:
using span cause view breaks avatar arrangement

chiragsalian committed:
bug fix

chiragsalian committed:
count correction

chiragsalian committed:
variable name correction

chiragsalian committed:
style overhaul to prevent js error and look good

chiragsalian committed:
fighting to get the styles right without js error

chiragsalian committed:
Merge branch 'main' into chirag-workspaces-list

chiragsalian commented:
I have no idea why PR Author Checklist is failing but we can ignore that for now.

sobitneupane reviewed (commented):
<img width="500" alt="Screen Shot 2022-10-21 at 08 21 54" src="https://user-images.githubusercontent.com/25876548/197098757-f01b1ff1-7d8d-40b1-aec5-a3561631ffdd.png">
Do we want to show "New Workspace" option in FAB if there are no workspace? Just making sure.

sobitneupane reviewed (commented):
While directly navigating to http://localhost:8080/settings/workspaces, "Create a new workspace" page is shown for a while before showing the list of workspaces.

https://user-images.githubusercontent.com/25876548/197101377-e61a926a-bb24-461d-a79e-30fce8136ea4.mov

sobitneupane reviewed (commented):
Red Dot icon not shown in "Workspaces" in Settings Page.

Steps to reproduce:
- Go offline
- Click on workspaces -> select a workspace -> manage members -> invite user +25857527444
- Close RHN
- Go back to online
- Click on profile icon to view Settings

https://user-images.githubusercontent.com/25876548/197103206-250251a1-25b0-495d-86ce-c4639a22103f.mov

sobitneupane reviewed (commented):
Avatar(with user uploaded image) of deleted workspace when offline defaults to avatar with person.
Steps To Reproduce:
- Go offline
- Delete a workspace
- Reload the page
- Notice avatar of deleted workspace in stack of avatars in "Workspace" in Settings
<img width="372" alt="Screen Shot 2022-10-21 at 09 28 15" src="https://user-images.githubusercontent.com/25876548/197106654-dfb4c835-cdc6-406b-9179-750028800d16.png">

sobitneupane reviewed (commented):
When there is only one workspace (not sure), Deleted Workspace in offline reappears in online. Might be related to delete functionality.

Steps to Reproduce:

- Create your first workspace
- Go offline
- Delete the workspace
- Go to workspace list page
- Go back online

https://user-images.githubusercontent.com/25876548/197108837-298a2d6b-8c8a-48e8-aedf-7baeccf411c9.mov

sobitneupane reviewed (commented):
In contrast to [above issue](https://github.com/Expensify/App/pull/11784#pullrequestreview-1150301544), deleted workspace in offline reappears in offline.

- Create two workspaces
- Go offline
- Delete one workspace
- Go back to online.
-  Go offline again

https://user-images.githubusercontent.com/25876548/197110484-4f6e444e-8858-4d5b-bdb1-6c24a0b7691e.mov

sobitneupane commented on src/components/MenuItem.js:
```diff
@@ -46,6 +47,8 @@ const defaultProps = {
     fallbackIcon: Expensicons.FallbackAvatar,
     brickRoadIndicator: '',
+    floatRightAvatars: [],
+    shouldStackHorizontally: false,
```
`shouldStackHorizontally` is not added in `propTypes`.

chiragsalian replied:
Updated

sobitneupane commented on src/pages/workspace/WorkspacesListPage.js:
```diff
@@ -0,0 +1,214 @@
+     * @returns {Number} the user wallet balance
+     */
+    getWalletBalance(isPaymentItem) {
+        return (isPaymentItem && Permissions.canUseWallet(this.props.betas))
```
Not subscribed to ONYXKEYS.BETAS to read betas.

chiragsalian replied:
Updated

JmillsExpensify commented:
> Do we want to show "New Workspace" option in FAB if there are no workspace? Just making sure.

For the time being yes. We'll move away from this logic further down in our product roadmap when we've implemented a more complete onboarding experience.

roryabraham reviewed (commented):
NAB but this seems like a really good component to add some storybook stories for

chiragsalian committed:
correcting red brick road dot logic

chiragsalian committed:
Merge branch 'main' into chirag-workspaces-list

chiragsalian committed:
updating logic for pending deleted workspaces

chiragsalian commented:
Updating this today. First priority is the App deploy and this is second on my list.

chiragsalian committed:
Merge branch 'main' into chirag-workspaces-list

chiragsalian committed:
logic correction

chiragsalian committed:
adding shouldStackHorizontally prop

chiragsalian committed:
subscribing to beta and adding prop

chiragsalian commented:
> While directly navigating to http://localhost:8080/settings/workspaces, "Create a new workspace" page is shown for a while before showing the list of workspaces.

I don't feel like this is a common use case since users could never go to `settings/workspaces` unless they manually set it that way. But either way i think we can make a follow up issue to polish this up if needed.

> Red Dot icon not shown in "Workspaces" in Settings Page.

Nice, great catch, i updated the code to correct this. Can you retest.

> Avatar(with user uploaded image) of deleted workspace when offline defaults to avatar with person.
Steps To Reproduce:

Updated code to not show pending deleted workspaces which would just automatically clean this up.

> When there is only one workspace (not sure), Deleted Workspace in offline reappears in online. Might be related to delete functionality.

This seems to work just fine for me.

> In contrast to https://github.com/Expensify/App/pull/11784#pullrequestreview-1150301544, deleted workspace in offline reappears in offline.

Odd issue, I'm not able to reproduce what you got and it works just fine for me. The difference in my testing is i usually turn off network per chrome tab via developer tools and not switch off the internet or my laptop. But either way, onyx data is usually how i expect so it seems to be working fine for me 🤷‍♂️ 

> NAB but this seems like a really good component to add some storybook stories for

I don't mind but I would prefer to create a follow-up issue for this to prevent this PR from growing in size.

Updated PR to address the remaining comments and retested. Ready for re-review.

chiragsalian committed:
lint fix

chiragsalian committed:
making lint happy

sobitneupane commented:
https://github.com/Expensify/App/pull/11784#pullrequestreview-1150257413 : I think We should create follow up issue.
https://github.com/Expensify/App/pull/11784#pullrequestreview-1150268507 : I cannot reproduce this issue any longer.
https://github.com/Expensify/App/pull/11784#pullrequestreview-1150287022 : No any avatar shown for deleted workspace. But we can find the deleted workspace in Workspaces list. I am not sure if it is an issue.

https://user-images.githubusercontent.com/25876548/197780671-410f9be0-e379-482e-98f3-5001945ef87e.mov


https://github.com/Expensify/App/pull/11784#pullrequestreview-1150301544 : I cannot constantly reproduce it. But I can still reproduce this issue once in 3-4 attempts. This time at first I see blank page without "Create a new workspace" message. After refreshing the page, I can still see the deleted workspace.

https://user-images.githubusercontent.com/25876548/197782927-a63af14c-40e8-4b3a-8c3d-63ae99d40b3f.mov

https://github.com/Expensify/App/pull/11784#pullrequestreview-1150305881 : I can still reproduce this issue. I followed the reproduction steps two times. On the second attempt, I was able to reproduce.

https://user-images.githubusercontent.com/25876548/197785154-2778281a-bbbc-4217-97c8-0ca2c0c12a0f.mov

chiragsalian commented:
https://github.com/Expensify/App/pull/11784#pullrequestreview-1150257413 - I agree. I created a follow-up [here](https://github.com/Expensify/App/issues/12128). I'll add the necessary labels and mark external once this code is merged.

https://github.com/Expensify/App/pull/11784#pullrequestreview-1150287022 - That's not an issue. In offline mode we should show the deleted workspace with a strikethrough. Unless you are online, hard to tell from your video. When online the deleted workspace should vanish.

https://github.com/Expensify/App/pull/11784#pullrequestreview-1150301544 - I tried the test 6 times in a row and could not reproduce. Also that issue does not look related to this PR. You should check the network request for that one because it looked like the deleted command failed. Because on refresh if you see the deleted workspace that means it was never deleted in the backend. So it doesn't look related to this PR.

https://github.com/Expensify/App/pull/11784#pullrequestreview-1150305881 - Also worked fine for me and i also feel this is the similar problem as previous where your network request is not really deleting the workspace which is why it shows up on reconnect. Not related to this PR since i don't change the behavior of those. 

Maybe try again by deleting node_modules and running npm i. Maybe some files are older or out of sync hence causing these problems? Either way, since i feel those are not related to this PR can we move ahead with reviewing or approving this PR. If those problems persist we can open follow-ups for them 🙂

sobitneupane reviewed (approved):
@chiragsalian QA steps are not added.
>  I verified the steps for Staging and/or Production testing are in the QA steps section

Changes look good and test well on all platforms.
🎀👀🎀 C+ reviewed

cc: @roryabraham @cristipaval 

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->

https://user-images.githubusercontent.com/25876548/197927955-3c1379b8-b595-4d3a-b045-4ad4d83023b3.mov


#### Mobile Web - Chrome
<!-- Insert screenshots of your changes on the web platform (from chrome mobile browser)-->

https://user-images.githubusercontent.com/25876548/197927992-2778624d-f2cb-4aa9-b0a0-f4786f539236.mov


#### Mobile Web - Safari
<!-- Insert screenshots of your changes on the web platform (from safari mobile browser)-->

https://user-images.githubusercontent.com/25876548/197928043-bd6f2f85-1f73-4113-a589-960554b95318.mov


#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->

https://user-images.githubusercontent.com/25876548/197927976-0836098a-537f-4879-b837-0763d48adb43.mov


#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

https://user-images.githubusercontent.com/25876548/197928054-8cfc4b5c-d7bc-4c60-9d10-5d33160ace96.mov


#### Android
<!-- Insert screenshots of your changes on the Android platform-->

https://user-images.githubusercontent.com/25876548/197928019-62944b58-61ca-4c17-bec7-bfe1677ecf99.mov

chiragsalian commented:
> QA steps are not added.

Oh hah, classic. Thanks Rory for updating the section to say Tests / QA steps.

btw @sobitneupane, for some of the issues you mentioned where workspaces are showing back up. Some other engineers reported the issue and we're discussing it here https://github.com/Expensify/App/issues/12109 and should have a fix for it soon.

chiragsalian commented:
I bumped @cristipaval on newDot to review this.

cristipaval reviewed (commented):
After deleting all workspaces, the placeholder in Workspaces is hidden.
Please see this recording. It occurs on desktop

https://user-images.githubusercontent.com/18078393/198380857-bec975fd-7750-472e-b1ff-e20a444a7cc6.mov

chiragsalian commented:
Works fine for me on desktop,

https://user-images.githubusercontent.com/14356145/198384115-20acbbab-6e69-4926-a7ae-80bd734737b9.mov

cristipaval reviewed (approved):
After deleting node_modules and running `npm i` again, it works well!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3342269560) to staging by @cristipaval in version: 1.2.21-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/3363432137) to production by @Julesssss in version: 1.2.21-4 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

arosiclair commented:
Hey the scenario for viewing a deleted/invalid workspace and backing out of it was missed when implementing this new page and caused #13276. Since this is a new page between the initial settings page and the workspace page, we should've updated the "not found" blocking view to just return you to the new list page instead of dismissing the modal entirely.

NikkiWines commented:
👋 This PR didn't optimistically filter out policies that were pending deletion resulting in https://github.com/Expensify/App/issues/13615. Since we want to optimistically assume success here, we should have removed the policies up front and added them back in the request failed.

francoisl commented:
👋 This PR caused [an issue](https://github.com/Expensify/App/issues/14564) when the app was configured in Spanish, because the empty Workspaces page didn't have correct horizontal padding. Not sure what could have been differently since the test steps don't say to test with the app in Spanish, but started a discussion [here](https://expensify.slack.com/archives/C049HHMV9SM/p1676402140771269) if anyone thinks we could have done something different to prevent the issue.

eVoloshchak commented:
👋 This PR was sorting workspaces alphabetically while not ignoring the case, which caused https://github.com/Expensify/App/issues/16925 and resolved by using
`.sortBy(policy => policy.title.toLowerCase())`
Not a regression, just something that was missed during the initial implementation

sobitneupane commented on src/components/MenuItem.js:
```diff
@@ -150,8 +153,18 @@ const MenuItem = (props) => {
+                        )}
                         {Boolean(props.brickRoadIndicator) && (
-                            <View style={[styles.alignItemsCenter, styles.justifyContentCenter]}>
+                            <View style={[styles.alignItemsCenter, styles.justifyContentCenter, styles.l4]}>
```
We had a small inconsistent padding issue caused by this change. 

Issue: https://github.com/Expensify/App/issues/18418
Issue Tittle: Workspaces icons and green dot has inconsistent padding
